tests/unit/s2n_connection_serialize_test.c

```diff
@@ -473,6 +500,488 @@ int main(int argc, char **argv)
         EXPECT_OK(s2n_send_and_recv_test(client_conn, new_server_conn));
     };
 
+    /* Self-talk: Test interaction between TLS1.2 session resumption and serialization */
+    {
+        DEFER_CLEANUP(struct s2n_config *resumption_config = s2n_config_new(), s2n_config_ptr_free);
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(resumption_config, chain_and_key));
+        EXPECT_SUCCESS(s2n_config_disable_x509_verification(resumption_config));
+        EXPECT_SUCCESS(s2n_config_set_serialized_connection_version(resumption_config, S2N_SERIALIZED_CONN_V1));
+        EXPECT_OK(s2n_resumption_test_ticket_key_setup(resumption_config));
+
+        /* Session tickets get sent during the handshake in TLS1.2, so they won't be available
+         * post-serialization. This test confirms that. */
+        {
+            struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+            EXPECT_NOT_NULL(client_conn);
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, resumption_config));
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, resumption_config));
+
+            DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+            EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+            EXPECT_OK(s2n_connections_set_io_stuffer_pair(client_conn, server_conn, &io_pair));
+
+            EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+            EXPECT_EQUAL(s2n_connection_get_actual_protocol_version(server_conn), S2N_TLS12);
+
+            /* Session ticket should have been received in the handshake */
+            EXPECT_TRUE(s2n_connection_get_session_length(client_conn) > 0);
+
+            /* Serialize and then free the client connection */
+            uint8_t buffer[S2N_SERIALIZED_CONN_TLS12_SIZE] = { 0 };
+            EXPECT_SUCCESS(s2n_connection_serialize(client_conn, buffer, sizeof(buffer)));
+
+            EXPECT_SUCCESS(s2n_connection_free(client_conn));
+
```

Comment:
We talked offline about trying to make this test more readable by restructuring it:
- Use a different context + DEFER_CLEANUP for each handshake to make the lifetimes clearer
- Fold this and the next test together by just adding another resumption handshake to this sequence of handshakes. So you'd have: initial handshake, resume from initial handshake, deserialized handshake, resume from deserialized handshake.